src/hit/api/evaluator/explicitdepthfinder.cpp

```diff
@@ -0,0 +1,147 @@
+         * meaning it starts at 0 and goes down. Thus, if the input ciphertext has he_level -1, we have
+         * already rescaled once after bootstrapping, and we are about to do so again. That means that
+         * the post_bootstrap_depth is (at least) 2 = 1 - (-1).
+         */
+        if (ct.bootstrapped()) {
+            scoped_lock lock(mutex_);
+            implicit_post_bootstrap_depth_ = max(implicit_post_bootstrap_depth_, 1 - ct.he_level());
+        }
+        // CT level is adjusted in CKKSEvaluator::rescale_metata_to_next
+    }
+
+    CKKSCiphertext ExplicitDepthFinder::bootstrap_internal(const CKKSCiphertext &ct, bool rescale_for_bootstrapping) {
+        // if rescale_for_bootstrapping, bootstrapping will implicitly consume one additional level to rescale the
+        // ciphertext first, ensure that if explict levels are set, we aren't already at level 0
+        if (rescale_for_bootstrapping && ct.he_level() == 0) {
+            LOG_AND_THROW_STREAM("Cannot rescale a level 0 ciphertext for bootstrapping");
+        }
+
+        // see comment in rescale_to_next_inplace_internal for explanation of arithmetic,
+        // and note that rescale_for_bootstrapping is either 0 or 1.
+        if (ct.bootstrapped()) {
+            scoped_lock lock(mutex_);
+            implicit_post_bootstrap_depth_ =
+                max(implicit_post_bootstrap_depth_, static_cast<int>(rescale_for_bootstrapping) - ct.he_level());
+        }
+        uses_bootstrapping = true;
+        // CT bootstrapped_ is adjusted in CKKSEvaluator::bootstrap
+        CKKSCiphertext bootstrapped_ct = ct;
+        bootstrapped_ct.he_level_ = 0;
+        return bootstrapped_ct;
+    }
+
+    int ExplicitDepthFinder::get_param_bootstrap_depth() const {
+        shared_lock lock(mutex_);
+
+        return max_contiguous_depth - get_param_eval_depth();
+    }
+
+    int ExplicitDepthFinder::get_param_eval_depth() const {
+        shared_lock lock(mutex_);
+
+        // max_contiguous_depth is set based on the maximum encryption level. Actual number of levels in the HE params
+        // may be more than this, i.e., this is a lower bound. explicit_post_bootstrap_depth_, if set, defines exactly
+        // how many post-boostrapping levels the parameters need.
+        // The implicit_post_bootstrap_depth_, which is based on the number of rescales post-bootstrapping and/or
+        // the level at which a bootstrapped ciphertext is re-bootstrapped, must be <=
+        // explicit_post_bootstrap_depth_
+
+        if (explicit_post_bootstrap_depth_ >= 0 && implicit_post_bootstrap_depth_ > explicit_post_bootstrap_depth_) {
+            LOG_AND_THROW_STREAM("explicit_post_bootstrap_depth_ < implicit_post_bootstrap_depth_: "
+                                 << explicit_post_bootstrap_depth_ << " < " << implicit_post_bootstrap_depth_);
+        }
+        return uses_bootstrapping ? max(implicit_post_bootstrap_depth_, explicit_post_bootstrap_depth_) : max_contiguous_depth;
```

Comment:
Is this lock used to manage the access to `implicit_post_bootstrap_depth_ `? But why `explicit_post_bootstrap_depth_ ` does not need a lock (e.g. in `set_explicit_post_bootstrap_depth`)?

Reply:
The lock is around the entire scope, which in this case is the entire function body. It's designed to protect _all_ internal state at once so that this evaluator will still produce the correct result on a multi-threaded circuit.

Reply:
Added some missing mutex guards; good catch.